drivers/api/no_os_spi.c

```diff
@@ -125,14 +132,21 @@ int32_t no_os_spi_transfer(struct no_os_spi_desc *desc,
 	if (desc->platform_ops->transfer)
 		return desc->platform_ops->transfer(desc, msgs, len);
 
+	no_os_mutex_lock(desc->mutex);
+
 	for (i = 0; i < len; i++) {
-		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff)
+		if (msgs[i].rx_buff != msgs[i].tx_buff || !msgs[i].tx_buff) {
+			no_os_mutex_unlock(desc->mutex);
 			return -EINVAL;
+		}
 		ret = no_os_spi_write_and_read(desc, msgs[i].rx_buff,
 					       msgs[i].bytes_number);
-		if (NO_OS_IS_ERR_VALUE(ret))
+		if (NO_OS_IS_ERR_VALUE(ret)) {
+			no_os_mutex_unlock(desc->mutex);
 			return ret;
+		}
 	}
 
+	no_os_mutex_unlock(desc->mutex);
 	return 0;
```

Comment:
this may be simplified by using `goto` statements as such:

```
for(...) {
ret = no_os_spi_write_and_read(desc, msgs[i].rx_buff,
					       msgs[i].bytes_number);
if (NO_OS_IS_ERR_VALUE(ret))
        goto out;
}

out:
        no_os_mutex_unlock(desc->mutex);

return ret;
```